### PR TITLE
Ignore unattached private version marker removals

### PR DESCRIPTION
**Pull request by napetrov**
## Summary
- ignore removed private ELF version-definition markers when no old exported symbol used that version node
- keep private version nodes with exported symbols as breaking via existing symbol-backed version-node detection
- add regression coverage for marker-only removals and private exported-symbol removals

## Real-world evidence
- Found in the recurring real-world validation campaign on conda-forge `libnsl` 2.0.0 -> 2.0.1.
- Before this fix, abicheck reported `BREAKING` solely for removed `LIBNSL_PRIVATE`.
- `abidiff` exited clean, dynamic export diff removed only `LIBNSL_PRIVATE`, and `readelf --dyn-syms` showed it was a zero-sized absolute version-definition marker rather than callable/data ABI.
- After this fix, the same comparison reports `COMPATIBLE_WITH_RISK` with only the new `GLIBC_2.14` requirement.

Artifacts are recorded in the local campaign report under `reports/abicheck-realworld-cron/artifacts/20260605-0247/`.

## Tests
- `pytest -q tests/test_sprint2_elf.py tests/test_diff_platform_deep.py::TestSymbolVersionChanges tests/test_elf_version_policy.py` (88 passed)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed unattached private symbol version markers from being reported as removed
  * Improved handling of deleted functions marked in debug info to avoid false breaking changes
  * Enhanced parameter rename detection to work correctly with incomplete header inputs
  * Refined ELF symbol filtering to exclude linker-internal boundary markers
  * Extended compiler-internal type detection to cover typedef aliases

* **Tests**
  * Added tests for symbol versioning edge cases and linker boundary marker filtering
  * Added tests for debug-info-only deleted function scenarios

### CODE REVIEW

File: abicheck/diff_platform.py [hunk 1]
```diff
@@ -482,6 +482,8 @@ def _diff_elf_symbol_versioning(old_elf: Any, new_elf: Any) -> list[Change]:
     old_def = set(old_elf.versions_defined)
     new_def = set(new_elf.versions_defined)
     for ver in sorted(old_def - new_def):
+        if _is_unattached_private_version_node(old_elf, ver):
+            continue
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/diff_platform.py:511]: Fixed in cafd351. `detect_version_script_missing` now discounts unattached private markers when computing `old_had_version_script` — so a version script whose only definition is `FOO_PRIVATE` (with no symbol bound to it) no longer counts as a real version script, and dropping it no longer re-emits `VERSION_SCRIPT_MISSING`. I hoisted the canonical `_is_unattached_private_version_node` helper into `diff_versioning` so the version-def-removal and version-script-missing paths share one definition. Added the marker-only regression test (`test_marker_only_private_version_script_removed_emits_no_warning`) that the previous test didn't cover.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/diff_symbols.py [hunk 2]
```diff
@@ -583,9 +584,12 @@ def _detect_newly_deleted_functions(
     produce spurious BREAKING findings.
     """
     changes: list[Change] = []
+    exported = exported_symbol_names(getattr(new_snapshot, "elf", None), FUNCTION_SYMBOL_TYPES)
     for mangled, f_new in new_all.items():
         if not f_new.is_deleted:
             continue
+        if f_new.deleted_from_dwarf and exported and mangled not in exported:
+            continue
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/diff_symbols.py]: Confirmed and fixed in 28ead8d. Reproduced: a new lib with an ELF symbol table that exports only data (no function symbols) caused a DWARF-only `DW_AT_deleted` internal member to be reported as `FUNC_DELETED_DWARF` → BREAKING, because the guard used `exported` truthiness (`... and exported and mangled not in exported`).

The guard now keys on ELF symbol-table presence (`has_elf_symbol_table = bool(new_elf.symbols)`) instead, so it tells "no ELF evidence available" apart from "ELF table present, function not exported". Verified three cases: (1) data-only exports → not reported; (2) no ELF table → falls back to visibility and still reports; (3) function actually exported+deleted → reported. Added `test_unexported_dwarf_deleted_not_reported_when_only_data_exported` and `test_dwarf_deleted_reported_when_no_elf_table`.

---
_Generated by [Claude Code](https://claude.ai/code)_
